File: src/VersionUp/NuspecVersionHandler.cs
namespace VersionUp;

using System;
using System.IO;
using System.Xml.Linq;

/// <summary>
/// Handles parsing and updating version elements in NuGet specification (.nuspec) files.
/// </summary>
public class NuspecVersionHandler : IVersionFileHandler
{
    /// <inheritdoc />
    public bool CanHandle(string filePath)
    {
        if (string.IsNullOrEmpty(filePath))
        {
            return false;
        }

        string extension = Path.GetExtension(filePath).ToLowerInvariant();

        return extension == ".nuspec";
    }

    /// <inheritdoc />
    public string? GetVersion(string fileContent)
    {
        if (string.IsNullOrWhiteSpace(fileContent))
        {
            return null;
        }

        try
        {
            XDocument doc = XDocument.Parse(fileContent);
            XElement? versionElement = FindVersionElement(doc);

            return versionElement?.Value;
        }
        catch
        {
            return null;
        }
    }

    /// <inheritdoc />
    public string UpdateVersion(string fileContent, string newVersion)
    {
        if (string.IsNullOrWhiteSpace(fileContent))
        {
            fileContent = "<package><metadata></metadata></package>";
        }

        XDocument doc = XDocument.Parse(fileContent);
        XElement? metadata = doc.Root?.Element("metadata");

        if (metadata == null)
        {
            metadata = new XElement("metadata");
            doc.Root?.Add(metadata);
        }

        XElement? versionElement = FindVersionElement(doc);

        if (versionElement == null)
        {
            versionElement = new XElement("version", newVersion);
            metadata.Add(versionElement);
        }
        else
        {
            versionElement.Value = newVersion;
        }

        string result = doc.ToString();

        return result;
    }

    /// <summary>
    /// Searches for the <c>&lt;version&gt;</c> element inside <c>&lt;metadata&gt;</c>.
    /// </summary>
    /// <param name="doc">The parsed XML document.</param>
    /// <returns>The version element, or <see langword="null"/> if not found.</returns>
    private static XElement? FindVersionElement(XDocument doc)
    {
        if (doc.Root == null)
        {
            return null;
        }

        XElement? metadata = doc.Root.Element("metadata");

        if (metadata != null)
        {
            XElement? version = metadata.Element("version");

            return version;
        }

        return null;
    }
}

File: vscode-extension/src/handlers/NuspecVersionHandler.ts
import * as path from 'path';
import { IVersionFileHandler } from './IVersionFileHandler';

/**
 * Handles parsing and updating version elements in NuGet specification (.nuspec) files.
 */
export class NuspecVersionHandler implements IVersionFileHandler {
    public canHandle(filePath: string): boolean {
        if (!filePath) {
            return false;
        }

        const extension = path.extname(filePath).toLowerCase();
        return extension === '.nuspec';
    }

    public getVersion(fileContent: string): string | null {
        if (!fileContent || !fileContent.trim()) {
            return null;
        }

        const versionMatch = fileContent.match(/<version>\s*([^<\s]+)\s*<\/version>/i);
        if (versionMatch) {
            return versionMatch[1];
        }

        return null;
    }

    public updateVersion(fileContent: string, newVersion: string): string {
        if (!fileContent || !fileContent.trim()) {
            return `<package>\n  <metadata>\n    <version>${newVersion}</version>\n  </metadata>\n</package>`;
        }

        if (/<version>\s*[^<\s]+\s*<\/version>/i.test(fileContent)) {
            return fileContent.replace(/(<version>\s*)[^<\s]+(\s*<\/version>)/i, `$1${newVersion}$2`);
        }

        const metadataMatch = fileContent.match(/<metadata>/i);
        if (metadataMatch && metadataMatch.index !== undefined) {
            const insertIndex = metadataMatch.index + metadataMatch[0].length;
            return fileContent.slice(0, insertIndex) + `\n    <version>${newVersion}</version>` + fileContent.slice(insertIndex);
        }

        const packageMatch = fileContent.match(/<package[^>]*>/i);
        if (packageMatch && packageMatch.index !== undefined) {
            const insertIndex = packageMatch.index + packageMatch[0].length;
            return fileContent.slice(0, insertIndex) + `\n  <metadata>\n    <version>${newVersion}</version>\n  </metadata>` + fileContent.slice(insertIndex);
        }

        return fileContent;
    }
}
